src/text/parsers/text_support.rs

```diff
@@ -22,6 +22,26 @@ pub(crate) enum StringFragment<'a> {
     EscapedNewline,
 }
 
+/// Checks the given input for a leading slash (`\`); if it finds one, it applies the provided
+/// `parser` to the text that follows. If the parser doesn't match, returns a fatal error.
+pub(crate) fn escape_and_then<'a, 'b, P>(
+    input: &'a str,
+    label: &'b str,
+    mut parser: P,
+) -> IonParseResult<'a, StringFragment<'a>>
+where
+    P: Parser<&'a str, StringFragment<'a>, IonParseError<'a>>,
+{
+    // If it doesn't start with a slash, it's not a match. Return a non-fatal error.
+    let (remaining, _slash) = char('\\')(input).upgrade()?;
+    // If the provided parser doesn't match what comes next, it's not a valid escape.
+    // Return a fatal error.
+    match parser.parse(remaining) {
+        Ok((remaining, string_fragment)) => Ok((remaining, string_fragment)),
+        Err(e) => fatal_parse_error(remaining, format!("could not parse {}: {}", label, e)),
+    }
+}
```

Comment:
:world_map: This helper function enforces that if the stream has a `\` in it, the thing that comes next must be a complete escape sequence. For example, `\u000B` is a valid unicode escape, but `\u00` is too short and should result in a fatal error.

Prior to this PR, if the escape sequence was incomplete it would be considered literal text instead. (`\u00` would appear in the string as `\00`.)